File: 0253-meeting-rooms-ii/0253-meeting-rooms-ii.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        required_rooms = 0
        pending_end = []
        intervals.sort()

        for start, end in intervals:
            while len(pending_end) and pending_end[0] <= start:
                heapq.heappop(pending_end)
            heapq.heappush(pending_end, end)
            required_rooms = max(required_rooms, len(pending_end))

        return required_rooms

File: 0253-meeting-rooms-ii/0253-meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        PriorityQueue<Integer> dueTime = new PriorityQueue<>();
        int rooms = 0;
        for (int i = 0; i < intervals.length; i++) {
            if (dueTime.size() == 0 || dueTime.peek() > intervals[i][0]) {
                rooms++;
            }
            else {
                dueTime.poll();
            }
            dueTime.add(intervals[i][1]);
        }

        return rooms;
    }
}